### PR TITLE
Draft: Fixed vulnerabilities in verify-install.sh

### PR DESCRIPTION
**Pull request by Kixunil**
The verification script contained three vulnerabilities:

* It executed the binary being verified - if it were malicious, it'd
  compromise the system of the user already
* The script didn't check if multiple signatures are actually from
  distinct people. A compromised developer could have duplicate their
  signature to fool the script into accepting a single signature.
* Even if executing the binary wasn't broken, no version checking
  was performed and thus downgrade attacks were possible.

This fixes all of them, but wasn't sure what to do with Docker-related
stuff.

**This is a breaking change** - it's impossible to do it without
breaking because the version must be specified externally to be secure.

#### Pull Request Checklist

This changes no Go code. Wanted to submit this ASAP, will get back later and go over checklist. Please LMK what to do about docker stuff meanwhile.

- [ ] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [ ] All changes are Go version 1.12 compliant
- [ ] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [ ] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [ ] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [ ] Any new logging statements use an appropriate subsystem and
  logging level
- [ ] Code has been formatted with `go fmt`
- [ ] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] New configuration flags have been added to `sample-lnd.conf`
- [ ] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [ ] Running `make check` does not fail any tests
- [ ] Running `go vet` does not report any issues
- [ ] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [ ] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

File: scripts/verify-install.sh [hunk 1]
```diff
@@ -158,29 +126,40 @@ for signature in $SIGNATURES; do
...
-  elif gpg --verify "$signature" 2>&1 | grep -q "No public key"; then
-    echo "Unable to verify signature $signature, no key available, skipping"
-    continue
+  VALID=0
+  for key in "${KEYS[@]}"; do
+    KEY_ID=$(echo $key | cut -d' ' -f1)
+    if GNUPGHOME="$TEMP_DIR/$KEY_ID" gpg --verify "$TEMP_DIR/$signature" "$TEMP_DIR/$MANIFEST" 2>&1 | grep -q "Good signature"; then
+      # We remove the key and signature to NOT count them more than once
+      # THIS IS CRUCIAL FOR VERIFICATION TO BE SECURE!!!
+      rm -rf "$TEMP_DIR/$KEY_ID" "$signature" || exit 1
```
Comment (guggero) [inline scripts/verify-install.sh:141]: Great idea to count the number of distinct keys.

**Review by guggero — COMMENTED** [review]
Thanks for looking into this!
I agree that executing a binary isn't the best idea if you haven't verified it yet.

I'm okay with this being a breaking change, as long as the PR also contains the update to the instructions on how to verify.
My main comment is about making the diff smaller by not making this any more generic than it needs to be.

File: scripts/verify-install.sh [hunk 2]
```diff
@@ -29,42 +29,22 @@ function check_command() {
...
-  fi
-  if [[ ! -f $LNCLI_BIN ]]; then
-    echo "ERROR: $LNCLI_BIN not found!"
-    exit 1
-  fi
-elif [[ $# -eq 0 ]]; then
-  # Make sure both binaries can be found and are executable.
-  check_command lnd
-  check_command lncli
+# If exactly three parameters are specified, we expect the first one to be lnd version and
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply scripts/verify-install.sh:32]: I personally prefer to verify `.tar.gz` before decompressing as that also avoids theoretical attack against decompressing code.

File: scripts/verify-install.sh [hunk 3]
```diff
@@ -29,42 +29,22 @@ function check_command() {
...
-    echo "ERROR: $LND_BIN not found!"
-    exit 1
-  fi
-  if [[ ! -f $LNCLI_BIN ]]; then
-    echo "ERROR: $LNCLI_BIN not found!"
-    exit 1
-  fi
-elif [[ $# -eq 0 ]]; then
-  # Make sure both binaries can be found and are executable.
-  check_command lnd
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply scripts/verify-install.sh:53]: Sounds foot-gunny, maybe I could add `--docker` switch to opt-into it instead?

File: scripts/verify-install.sh [hunk 4]
```diff
@@ -75,62 +55,51 @@ if ! command -v "$SHA_CMD"; then
...
-  echo "ERROR: Invalid hash for lnd: $LND_SUM!"
-  exit 1
-fi
-if [[ ${#LNCLI_SUM} -ne 64 ]]; then
-  echo "ERROR: Invalid hash for lncli: $LNCLI_SUM!"
+  echo "ERROR: Invalid version of lnd specified: $LND_VERSION"
   exit 1
 fi
 
+# FIXME: how to fix this?
```
[the remark this answers is not in the document]
Comment (Kixunil) [reply scripts/verify-install.sh:65]: Maybe combine with the `--docker` switch suggested above?

File: scripts/verify-install.sh (hunk 1 above)
Comment (Kixunil) [reply scripts/verify-install.sh:141]: I'm glad you like it, it didn't felt entirely clean but hopefully least foot-gunny.

**Comment by cfromknecht** [comment]
@Kixunil thanks for a thorough review of the verification, these are definitely issues we want to address to improve the robustness of the script.

From my reading, passing in the explicit version as an argument seems to be the most crucial of these fixes so that we don't inadvertently execute malicious binaries on the host system before verifying. I've put up an rc7 with just this patch #5051 so that this can be included in 0.12.1. The other two I think we can look towards making iterative improvements in subsequent releases if that's okay with you?

**Comment by Kixunil** [comment]
@cfromknecht Sure, that makes sense, I will rebase my PR. The "multisig" approach is essentially broken without the fix but it's no worse than verifying a single signature before. It should be mentioned at appropriate place(s) that the script does 1 of N verification, so people don't mistakenly rely on it being M of N.

**Comment by Kixunil** [comment]
Replaced by #5053

**Pull request closed without merging**